sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp

```diff
@@ -530,211 +376,99 @@ class wi_element<uint16_t, NumRows, NumCols, Layout, Group> {
-    M.spvm = __spirv_VectorInsertDynamic(
-        M.spvm,
-        make_bf16(make_fp32(__spirv_VectorExtractDynamic(M.spvm, idx)) +
-                  make_fp32(rhs)),
-        idx);
-    return *this;
-#else
-    (void)rhs;
-    throw runtime_error("joint matrix is not supported on host device.",
-                        PI_INVALID_DEVICE);
-#endif // __SYCL_DEVICE_ONLY__
-  }
-
-  friend uint16_t
-  operator-(const wi_element<uint16_t, NumRows, NumCols, Layout, Group> &lhs,
-            const uint16_t &rhs) {
-#ifdef __SYCL_DEVICE_ONLY__
-    return make_bf16(
-        make_fp32(__spirv_VectorExtractDynamic(lhs.M.spvm, lhs.idx)) -
-        make_fp32(rhs));
-#else
-    (void)lhs;
-    (void)rhs;
-    throw runtime_error("joint matrix is not supported on host device.",
-                        PI_INVALID_DEVICE);
+#if __SYCL_DEVICE_ONLY__
+#define OP(opassign, op)                                                       \
+  wi_element &operator opassign(const uint16_t &rhs) {                         \
+    M.spvm = __spirv_VectorInsertDynamic(                                      \
+        M.spvm,                                                                \
+        make_bf16(make_fp32(__spirv_VectorExtractDynamic(M.spvm, idx)          \
+                                op make_fp32(rhs))),                           \
+        idx);                                                                  \
+    return *this;                                                              \
+  }
+#else // __SYCL_DEVICE_ONLY__
+#define OP(opassign, op)                                                       \
+  wi_element &operator opassign(const uint16_t &rhs) {                         \
+    (void)rhs;                                                                 \
+    throw runtime_error("joint matrix is not supported on host device.",       \
+                        PI_INVALID_DEVICE);                                    \
+  }
+#endif // __SYCL_DEVICE_ONLY__
+  OP(+=, +)
+  OP(-=, -)
+  OP(*=, *)
+  OP(/=, /)
+#undef OP
+
+#if __SYCL_DEVICE_ONLY__
+#define OP(type, op)                                                           \
+  friend type operator op(                                                     \
+      const wi_element<uint16_t, NumRows, NumCols, Layout, Group> &lhs,        \
+      const uint16_t &rhs) {                                                   \
+    return make_bf16(make_fp32(                                                \
+        __spirv_VectorExtractDynamic(lhs.M.spvm, lhs.idx)) op make_fp32(rhs)); \
+  }                                                                            \
+  friend type operator op(                                                     \
+      const uint16_t &lhs,                                                     \
+      const wi_element<uint16_t, NumRows, NumCols, Layout, Group> &rhs) {      \
+    return make_bf16(make_fp32(                                                \
```

Comment:
but this converter function needs two template parameters, one of which for 'type' and another for input's type. if we specialize for uint16_t, that means we do partial specialization for function templates and it is not allowed in c++.
template <typename T1, typename T2> T2 converter(const T1 &rhs) {
  return static_cast<T2>(rhs);
}

Reply:
I see. You could do it by wrapping it in a `struct`, i.e. something like
```c++
template <typename T1, typename T2> struct Converter {
    static T2 convert(const T1 &from) {
        return static_cast<T2>(from);
    }
};

template <typename T> struct Converter<uint16_t, T> {
    static T convert(const uint16_t &from) {
        return make_bf16(from);
    }
};
```
and then call `Converter<T1, T2>::convert(x)`. It does make it a little verbose so I am not sure how much benefit there is to it.

Reply:
it works for me and it seems we need another parameter in macro fumction, as shown in https://github.com/intel/llvm/pull/5532/commits/2bc6c8c1e57778b54cc83fe54942e37e364001b9

Reply:
Will `make_f32` not always result in a float? From what I can tell it should be fine to just replace `input_type` with `float`.

Reply:
relop return a bool value while binop return a float, if we need to unify the relop and binop, another parameter 'input_type' is needed.

Reply:
The return type of the operator is `type`, right? `input_type` is the type given to the converter to convert to `type`. In both cases, the converter is given the output of `make_fp32` which will always be a float. What you're currently doing is telling the converter to convert a `bool` to a `bool`, which it can do even though you pass it a float because of implicit conversion of float to bool.

Reply:
>>>>>In both cases, the converter is given the output of make_fp32 which will always be a float.
This is not correct i think.
if it is relop, make_fp32(__spirv_VectorExtractDynamic(lhs.M.spvm, lhs.idx)) op make_fp32(rhs) return a bool. the converter is given a bool.
if it is binop, make_fp32(__spirv_VectorExtractDynamic(lhs.M.spvm, lhs.idx)) op make_fp32(rhs) return a float. the converter is given a float.

Reply:
Apologies, you are absolutely right. I completely forgot that the operator was the last operation before the converter. I'll let you decide which version you prefer.

Reply:
Anyway unifying these two operations is quite brilliant! i think let us keep current code?